### ISSUE TITLE
Info/Help: Cannot get meson to detect and use Visual Studio Code under GitHub Actions

### ISSUE THREAD

**Issue by Framstag** [mesonbuild/meson#8638]
**Describe the bug**
I'm currently trying to get a C++ meson based build running under GitHub Actions. I already got meson to build under other CI platforms, this is not a principle problem, but more a problem of search paths etc...

I'm aware of https://mesonbuild.com/Using-with-Visual-Studio.html (up to now I used the vs2019 backend, but in the migration I'll to do not use this).

I'm using this GitHub action: https://github.com/Framstag/libosmscout/blob/vs_github_action/.github/workflows/build_and%20test_on_vs2019_meson.yml

Example of a working build under Appveyor: https://ci.appveyor.com/project/Framstag/libosmscout/builds/38643530/job/1tkvuoinmdplm6sv

The appveyor build does a call to `call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvarsall.bat" amd64`, which the GitHub build currently does not, since I trust the msbuild setup action of microsoft ;-) Making this call would be my next try...

Running it results in the following output: https://github.com/Framstag/libosmscout/actions/runs/737619236

**To Reproduce**
(meson build file: https://github.com/Framstag/libosmscout/blob/master/meson.build, but IMHO it is not important)

See above. 

**Expected behavior**
build runs.

**system parameters**
See above, especially output in the build for environment and content of path variable.

This is (very) likely not a bug in meson itself, I raised the issue to get some background what exactly meson expects in PATH and possibly other environments variables to successfully run a Visual Studio based builds. This would allow me to "do the right thing". If it is calling vcvarsall.bat, than this is OK - but please add it to the documentation ;-)

I thus more so want you to make you aware of the situation to possibly add details to existing documentation.

I'll try to now call vcvarsall.bat and will give you feedback.

**Comment by Framstag** [comment]
I'm lost now.

I'm now calling `call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvars64.bat"` explicitly. I'm also setting VSINSTALLDIR (C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\) since this is something that meson seems to be missing (there was a hint printed during one of my tries).

If I'm enforcing Visual Studio detection with `--backend vs` This results in internal debug messages and in the end gcc 8 detection (MSYS, which is also installed in the GitHub VM), but no real errors. meson though exists with an (unspecific) error.
https://github.com/Framstag/libosmscout/runs/2318445194?check_suite_focus=true

If I drop `--backend vs`the build runs, but with the also installed gcc 8 (MSYS)

It looks like Visual Studio is not correctly detected, but I do not see what is missing. I need help.

**Comment by dvdhrm** [comment]
GitHub "recently" fixed some of their environment propagation in github-actions. Are you sure calling `vcvars64.bat` will have effect on following steps?

Btw., I wrote a wrapper around "MSVC++/DevShell.dll" which works fine for me: https://dvdhrm.github.io/2021/04/21/meson-msvc-github-actions/

**Comment by Framstag** [comment]
> Are you sure calling vcvars64.bat will have effect on following steps?

No. I recently did check the command sequence on my local system regarding environment changes and compared it to the debug output in above mentioned builds. There are definitely differences that make me believe this might be a GitHub Actions problem. I'll read your article :-)

**Comment by Framstag** [comment]
Fine article. I feared that I have to set a number of environment variables to emulate enough Dev Shell to get meson to work. That was part of my initial request,to find out, what meson exactly needs.

Cool, that you already have found out a little more robust approach (though I question myself, if you use offical APIs).

As you hint at possibly coming better support in meson (though no actual code yet): I could live with passing the compiler installation directory and (optionally) version to meson and meson getting the rest to work on its own.

I'll try out your action in the coming days as soon as I have some time.

Many thanks!
